Add EvolveNewtonianEuler Executable

carmaza opened the pull request:
## Proposed changes

The order is

- Add missing members to `NewtonianEuler` system.
- Add NewtonianEuler Initializer. 
  This commit adds two "new" files that should be replaced by something that doesn't duplicate code:
  + `src/Evolution/Conservative/UpdatePrimitivesNoRecovery.hpp`, which should be deleted in favor of making `UpdatePrimitives.hpp`handle systems without a list of prim from cons recovery methods.
  + `src/Evolution/Initialization/InterfaceNewtonianEuler.hpp`, which should be deleted in favor of making  `Interface.hpp` handle systems without GR variables, as well as making it capable of initializing system-specific compute items on the interface.

  In addition, `src/Evolution/Systems/NewtonianEuler/Initialize.hpp`  is also mostly a copy-paste of other systems, so we should somehow factor that out too (maybe in a future PR in order to deal with all systems at once).
- Add EvolveNewtonianEuler Executable.
 (2-3D for now because there is no 1D initial data in develop yet!!!)

Any suggestions to deal with the code duplication in the initialization would be appreciated!

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

carmaza committed:
Add missing members to NewtonianEuler system

carmaza committed:
Add NewtonianEuler Executable

kidder reviewed (commented):
this test gives output which needs to either turned off or removed (see Poisson executable)

Otherwise, the last 3 commits look good (group review with @nilsdeppe @wthrowe )

carmaza commented:
I just removed the options of `EventsAndTriggers` in the input files, as done in other Evolution systems.

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/System.hpp:
```diff
@@ -5,15 +5,62 @@
+#include "PointwiseFunctions/Hydro/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Tags {
```
`cond` and `endcond` around this please.

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/System.hpp:
```diff
@@ -5,15 +5,62 @@
+  static constexpr size_t thermodynamic_dim =
+      EquationOfStateType::thermodynamic_dim;
+
+  // Compute item for pressure not currently implemented in SpECTRE,
```
Why isn't a compute item added? Why is this different from GRMHD?

carmaza replied:
If I'm not wrong, one would like to add the pressure as a generic compute item, where the function to compute it is retrieved from the equation of state. What is done in this system *is* essentially the same that was done in GRMHD. 

If we implement the compute item, we might also want to do it for all hydro systems at once. Definitely something for a different PR.

nilsdeppe replied:
I think the reason we can't currently have a compute item is we need PDALs to get const global cache things into the DataBox. Future problem it is :)

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Initialize.hpp:
```diff
@@ -0,0 +1,166 @@
+}  // namespace detail
+
+template <size_t Dim>
+struct Initialize {
```
Can you please file an issue to remove this once PDALs are in? This code is almost identical to the GRMHD case and I think we should figure out how to not do that once we have PDALs :)

carmaza replied:
Sure, I'll also mention that it should be removed from GRMHD (and possibly from GH, if it gets merged soon).

nilsdeppe replied:
Thanks!

nilsdeppe commented on src/Evolution/Initialization/InterfaceNewtonianEuler.hpp:
```diff
@@ -0,0 +1,103 @@
+// Distributed under the MIT License.
```
Two points on this:

- Why is this necessary? It should be clearly stated in the documentation why the current generic one isn't used.
- This should be in the NE directory since it's system specific.

nilsdeppe commented on src/Evolution/Initialization/InterfaceNewtonianEuler.hpp:
```diff
@@ -0,0 +1,103 @@
+/// - Removes: nothing
+/// - Modifies: nothing
+template <typename System>
+struct Interface {
```
Just make the generic one allow you to add compute face and ext tags, and while it's being changed it also should only add prims if the system has prims.

carmaza replied:
Yes, that's what I wanted too, but I'm afraid I'll need a more detailed explanation on how to do this :) Otherwise, I'll move this file to the NE directory.

nilsdeppe replied:
Okay, maybe we can talk about this in person tomorrow (Wednesday)?

nilsdeppe replied:
Here are the metafunctions and code that allows you to pass in a typelist of tags where the first tag will be substituted in with the `Direction` tag.
```cpp
template <typename DirectionsTag, typename VarsTag>
struct Slice {};

template <typename T>
struct SoundSpeed {};

template <template <class...> class F, typename... Ts>
struct interface_holder {
  template <typename Dir>
  using f = F<Dir, Ts...>;
};

template <typename InterfaceHolder, typename Direction>
struct add_direction_tag {
  using type = typename InterfaceHolder::template f<Direction>;
};

using tags = tmpl::list<interface_holder<Slice, SoundSpeed<double>>>;

struct Dir {};

using result = tmpl::transform<tags, add_direction_tag<tmpl::_1, Dir>>;

td<result> aoeu;
```
The compiler output is:
```
test.cpp:167:12: error: implicit instantiation of undefined template
      'td<brigand::list<Slice<Dir, SoundSpeed<double> > > >'
td<result> aoeu;
```
which shows that the transformation was done correctly. The `interface_holder` class can probably just go into the `Initialization` namespace, while the metafunction `add_direction_tag` can go inside the `Initialization::Interface` class.

carmaza commented:
Rebased on develop. Updated commit list is:

- Add missing members to system
- Add initialization:
  + I no longer need to add an "UpdatePrimitivesNoRecovery" file, as `MutateApply` can handle that in the executable (although I needed to add an extra fixup commit to make it work; see below). 
  + Initialization for the system is still a copy-paste of other systems. I can file an issue to remove that (as originally requested), or let it be for now and deal with that after pdals are in (not sure how but Im open to learn).
  + Initialization of the interface is specific for this system, and it's a copy-paste from the generic Interface.hpp file, because I need to add compute items specific to this system. (I did manage to make the generic file handle primitives and spacetime variables as optional arguments, though. Maybe that could be added in a different PR?)
- Fixup commit to add a template paratemer to `MutateApply` (Was this class actually tested? I can't find anywhere in SpECTRE where it's being used.)

- Add EvolveNewtonianEuler executable with 2 and 3D input files. 1D coming soon along with a 1D analytic solution (Riemann problem is the best candidate.)

carmaza committed:
Add ComputeTags Initialization

carmaza committed:
Fix ctest label for ParallelAlgorithms tests

carmaza commented:
Rebased on develop. The commit order is :

- add missing members to system
- fix ctest labels in the recently moved ParallelAlgorithms tests (let me know if this isn't needed)
- add support for initializing ComputeTags
- add NewtonianEuler executable updated to use PDALs.

nilsdeppe commented on src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp:
```diff
@@ -0,0 +1,263 @@
+
+              Parallel::PhaseActions<
+                  Phase, Phase::RegisterWithObserver,
+                  tmpl::list<Actions::AdvanceTime,
```
With #1604 merged this `AdvanceTime` can be removed

nilsdeppe commented on src/ParallelAlgorithms/Initialization/AddComputeTags.hpp:
```diff
@@ -0,0 +1,39 @@
+
+namespace Initialization {
+namespace Actions {
+template <typename ComputeTagsList>
```
Could you please write dox for this following the format that other actions use? It looks like the other action in this directory is in the `ActionsGroup`.

carmaza replied:
so should this header go in `ParallelAlgorithms/Initialization/Actions` then?

nilsdeppe replied:
hmm yea that seems most consistent

carmaza replied:
Changed the location and added some dox. Let me know if something's wrong or if there are better words for it.

nilsdeppe commented on tests/Unit/ParallelAlgorithms/Initialization/Test_AddComputeTags.cpp:
```diff
@@ -0,0 +1,187 @@
+#include <tuple>
+#include <utility>
+
+#include <iostream>
```
Whoops, I think this stuck around from debugging and can be removed now :)

nilsdeppe reviewed (commented):
Just a couple minor comments. I'm happy for you to squash things immediately.

This is really nice. Good job! :)

fmahebert commented on src/Evolution/Executables/NewtonianEuler/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+  MathFunctions
+  NewtonianEuler
+  NewtonianEulerSolutions
+  Limiters
```
Alphabetize

fmahebert commented on src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp:
```diff
@@ -0,0 +1,262 @@
+                 OptionTags::EventsAndTriggers<events, triggers>>;
+
+  static constexpr OptionString help{
+      "Evolve the Newtonian Euler system in consevative form.\n\n"};
```
Typo in "consevative"

fmahebert reviewed (commented):
Two minor observations, I don't mind if you prefer to fix these in a subsequent PR 🙂

carmaza commented:
thanks @fmahebert they should be fixed now.

fmahebert reviewed (approved):
Speaking just to the Newtonian Euler part (not the ComputeTag init part): LGTM
